man/lsinitrd.1.adoc
= LSINITRD(1)
:doctype: manpage
:man source:   dracut
:man manual:   dracut
:man version:  {version}

NAME
----
lsinitrd - tool to show the contents of an initramfs image

SYNOPSIS
--------
*lsinitrd* ['OPTION...'] [<image> [<filename> [<filename> [...] ]]]

*lsinitrd* ['OPTION...'] -k <kernel version>

DESCRIPTION
-----------
lsinitrd shows the contents of an initramfs image. if <image> is omitted, then
lsinitrd determines the default location based on the local configuration
or Linux distribution policy.

OPTIONS
-------
**-h, --help**::
    print a help message and exit.

**-s, --size**::
    sort the contents of the initramfs by size.

**-f, --file** _<filename>_::
    print the contents of <filename>.

**-k, --kver** _<kernel version>_::
    inspect the initramfs of <kernel version>.

**-m, --mod**::
    list dracut modules included of the initramfs image.

**--unpack**::
    unpack the initramfs to the current directory, instead of displaying the contents.
    If optional filenames are given, will only unpack specified files, else the whole image will be unpacked.
    Won't unpack anything from early cpio part.

**--unpackearly**::
    unpack the early microcode initramfs to the current directory, instead of displaying the contents.
    Same as --unpack, but only unpack files from early cpio part.

**-v, --verbose**::
    unpack verbosely

ENVIRONMENT
-----------

_DRACUT_BINDIR_::
    overrides the directory containing Dracut executables,
    e.g. **skipcpio**. If running **lsinitrd** installed in a cross-compiled
    sysroot, use this option to provide host Dracut binaries.
+
Default:
    _$prefix/lib/dracut_

AVAILABILITY
------------
The lsinitrd command is part of the dracut package and is available from
link:$$https://github.com/dracut-ng/dracut$$[https://github.com/dracut-ng/dracut]

AUTHORS
-------
Harald Hoyer

Amerigo Wang

Nikoli

SEE ALSO
--------
man:dracut[8]


NEWS.md
[Rendered view](https://github.com/dracut-ng/dracut/blob/main/NEWS.md)

dracut-112
==========

#### Notes

This release includes the following security related commits:
* **network-legacy:**  sanitize DHCP values in dhclient-script.sh ([11577739](https://github.com/dracut-ng/dracut/commit/11577739221ff38c1fd29abbba51a6c797376ed6))
* **overlayfs-crypt:**  prevent eval injection in parse_overlay_opts() ([584e95be](https://github.com/dracut-ng/dracut/commit/584e95bea90cedd12d9190e9b09bb45e1aa26b08))
* **dracut-lib:**  sanitize variable assignments using eval ([4d2d812b](https://github.com/dracut-ng/dracut/commit/4d2d812bed056ca2c671612d960911b075c99dd9))
* **base:**  add escape function implementing printf %q ([207c3397](https://github.com/dracut-ng/dracut/commit/207c339728eff81127469c2f6fe106447c781009))
* **systemd-networkd:** escape values from DHCP options ([b456aee2](https://github.com/dracut-ng/dracut/commit/b456aee2fc61c47db0d9a408634820969e7359df))
* **iscsi:**  normalize initiator and target names ([843af3e7](https://github.com/dracut-ng/dracut/commit/843af3e751782285cd2bc7c9a3a9d33e843fb91d))
* **net-lib:**  validate iSCSI LUN parameters ([05c02db9](https://github.com/dracut-ng/dracut/commit/05c02db994ca322371a40695a3cc223faa1c7258))

This release includes the following new dracut modules:
* **chrony:**  introducing the chrony module ([4d64d0b9](https://github.com/dracut-ng/dracut/commit/4d64d0b99dd7267764272dbc4a210ccbda2db3db))
* **qcom-adsp:**  add module to load Qualcomm ADSP module pre-udev ([647a685d](https://github.com/dracut-ng/dracut/commit/647a685d3567dcf42fde10e1ae258967441e3223))

#### Bug Fixes

*   remove basename calls ([1898470b](https://github.com/dracut-ng/dracut/commit/1898470b01f6ba8b2d3660ef438eae37e362c8de))
*   remove dirname calls ([66111f49](https://github.com/dracut-ng/dracut/commit/66111f4969f246995686b2ef67d851282629ed5d))
*   replace true by : ([37717095](https://github.com/dracut-ng/dracut/commit/37717095083ebdf46fce7c7c399b6713cd53f232))
* **base:**  use printf instead of echo for hook variable ([ceadcea0](https://github.com/dracut-ng/dracut/commit/ceadcea055fead6b8a218c5e854b68ff3e0ac603))
* **bluetooth:**  skip warning when module is explicitly requested ([f334ea7c](https://github.com/dracut-ng/dracut/commit/f334ea7cf7a95c042a47942039ad54fbcdf46c39))
* **crypt:**  do not call return to exit a systemd generator ([fbb7270e](https://github.com/dracut-ng/dracut/commit/fbb7270edcf9fbc1c60959dc35e532dfe5c382b8), closes [#2490](https://github.com/dracut-ng/dracut/issues/2490))
* **dbus:**  restrict After/Requires removal patterns ([95a4da1b](https://github.com/dracut-ng/dracut/commit/95a4da1b2273b29dd4d558790b2f301e46616040))
* **docs:**  declare the correct location to report vulnerabilities ([6a9d409f](https://github.com/dracut-ng/dracut/commit/6a9d409fe418439319377b90ab5717b12d52915f))
* **dracut:**
  *  write force_drivers cmdline file in one step ([f72f94a2](https://github.com/dracut-ng/dracut/commit/f72f94a25dcf125fc050250159ad5efe6831d088))
  *  use dstdir in build_ld_cache() ([6b7db010](https://github.com/dracut-ng/dracut/commit/6b7db010224221af4a09772b73a40e759ad30506))
  *  do not exit the loop in push_user_devs() without checking all items ([573b91ee](https://github.com/dracut-ng/dracut/commit/573b91ee99bb1890f89f073962cb96680379acee))
  *  do not exit the loop in push_host_devs() without checking all items ([b57d2c2e](https://github.com/dracut-ng/dracut/commit/b57d2c2e6f5550b2047be74bf6cfe82a5a3f880a))
  *  log warning if touch fails ([fcffd4a0](https://github.com/dracut-ng/dracut/commit/fcffd4a0e3ee840b903c4ab45d06eb1ca881cce3))
  *  two typos in the 3cpio fallback ([852aeeb2](https://github.com/dracut-ng/dracut/commit/852aeeb21f6771a4112f0d08061ac19518dd149b))
  *  default to hostonly_cmdline=no in chroot ([67d82872](https://github.com/dracut-ng/dracut/commit/67d82872bb0e2bec7d046cce04de939edd30f453))
  *  properly pass user-supplied SBAT to ukify ([abef6535](https://github.com/dracut-ng/dracut/commit/abef65354833834092df092c83d0cbcbe8b85508))
  *  remove support for tmpfilesconfdir ([fb09ec26](https://github.com/dracut-ng/dracut/commit/fb09ec26b124705028c371089e42608c4e1585de))
  *  support chmod applet from busybox ([8a0ee588](https://github.com/dracut-ng/dracut/commit/8a0ee588c148a5cfc8b5f58ae7716eedde840562))
* **dracut-functions:**
  *  reset _found flag per rule in inst_rules() ([a0383c8d](https://github.com/dracut-ng/dracut/commit/a0383c8d1260745fa16c7249ac04310e0473fb47))
  *  correct hostonly path for Gentoo sysusers ([f037de42](https://github.com/dracut-ng/dracut/commit/f037de42be8ca69ef0fe422ae50b699c131754a1))
* **dracut-install:**
  *  check return value in modalias read() ([379e6b35](https://github.com/dracut-ng/dracut/commit/379e6b359d8c47b86e0359d85d7c9ccbac5b8d88))
  *  handle empty string in dir_len() ([0224b30e](https://github.com/dracut-ng/dracut/commit/0224b30e43020639095d3c98df073e297ba233a3))
  *  remove FTS_NOSTAT in install_modules() fts traversal ([40e7af01](https://github.com/dracut-ng/dracut/commit/40e7af014b47ab91ef559610af3cb3f8dbf0fbb9))
* **dracut-lib:**  sanitize variable assignments using eval ([4d2d812b](https://github.com/dracut-ng/dracut/commit/4d2d812bed056ca2c671612d960911b075c99dd9))
* **dracut-systemd:**  add ordering constraints ([8aa5c93b](https://github.com/dracut-ng/dracut/commit/8aa5c93b29f1bbd214d257e169489b752ab71074))
* **dracut.conf.5:**  move fstab/chroot warning to hostonly_mode section ([1374dc95](https://github.com/dracut-ng/dracut/commit/1374dc954ebd7ffee223932bcef23060e7941109))
* **iscsi:**
  *  normalize initiator and target names ([843af3e7](https://github.com/dracut-ng/dracut/commit/843af3e751782285cd2bc7c9a3a9d33e843fb91d))
  *  handle empty URI in firmware boot mode ([3c1a3846](https://github.com/dracut-ng/dracut/commit/3c1a3846a0435cc553494d4e7a40258ce3329cb6))
* **kernel-modules:**

lsinitrd.sh
#!/bin/bash
#
# Copyright 2005-2010 Harald Hoyer <harald@redhat.com>
# Copyright 2005-2010 Red Hat, Inc.  All rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

usage() {
    {
        echo "Usage: ${0##*/} [options] [<initramfs file> [<filename> [<filename> [...] ]]]"
        echo "Usage: ${0##*/} [options] -k <kernel version>"
        echo
        echo "-h, --help                  print a help message and exit."
        echo "-s, --size                  sort the contents of the initramfs by size."
        echo "-m, --mod                   list modules."
        echo "-f, --file <filename>       print the contents of <filename>."
        echo "--unpack                    unpack the initramfs, instead of displaying the contents."
        echo "                            If optional filenames are given, will only unpack specified files,"
        echo "                            else the whole image will be unpacked. Won't unpack anything from early cpio part."
        echo "--unpackearly               unpack the early microcode part of the initramfs."
        echo "                            Same as --unpack, but only unpack files from early cpio part."
        echo "-v, --verbose               unpack verbosely."
        echo "-k, --kver <kernel version> inspect the initramfs of <kernel version>."
        echo
    } >&2
}

[[ $dracutbasedir ]] || dracutbasedir=/usr/lib/dracut

sorted=0
modules=0
unset verbose
declare -A filenames

unset POSIXLY_CORRECT
TEMP=$(getopt \
    -o "vshmf:k:" \
    --long kver: \
    --long file: \
    --long mod \
    --long help \
    --long size \
    --long unpack \
    --long unpackearly \
    --long verbose \
    -- "$@")

# shellcheck disable=SC2181
if (($? != 0)); then
    usage
    exit 1
fi

eval set -- "$TEMP"

while (($# > 0)); do
    case $1 in
        -k | --kver)
            KERNEL_VERSION="$2"
            shift
            ;;
        -f | --file)
            filenames[${2#/}]=1
            shift
            ;;
        -s | --size) sorted=1 ;;
        -h | --help)
            usage
            exit 0
            ;;
        -m | --mod) modules=1 ;;
        -v | --verbose) verbose="--verbose" ;;
        --unpack) unpack=1 ;;
        --unpackearly) unpackearly=1 ;;
        --)
            shift
            break
            ;;
        *)
            usage
            exit 1
            ;;
    esac
    shift
done

EXTRACTOR=
if command -v 3cpio > /dev/null; then
    if threecpio_help_output=$(3cpio --help); then
        if [[ $threecpio_help_output == *--make-directories* ]]; then
            EXTRACTOR=3cpio
        fi
    elif command -v cpio > /dev/null; then
        echo "Warning: Calling '3cpio --help' failed. Cannot check if 3cpio supports --make-directories. Falling back to cpio."
    else
        echo "Warning: Calling '3cpio --help' failed. Cannot check if 3cpio supports --make-directories."
        EXTRACTOR=3cpio
    fi
    unset threecpio_help_output
fi
if ! [[ $EXTRACTOR ]]; then
    if [[ -f "$dracutbasedir/src/extractinitrd/extractinitrd" ]]; then
        EXTRACTOR="$dracutbasedir/src/extractinitrd/extractinitrd"
    else
        EXTRACTOR="$dracutbasedir/extractinitrd"
    fi
    if ! [[ -x $EXTRACTOR ]]; then
        echo
        echo "Error: '$EXTRACTOR' not found, cannot continue!" >&2
        echo
        exit 1
    fi
fi

if ! [[ $KERNEL_VERSION ]]; then
    if type -P systemd-detect-virt &> /dev/null && ! systemd-detect-virt -c &> /dev/null && ! systemd-detect-virt -r &> /dev/null; then
        KERNEL_VERSION="$(uname -r)"
    else
        # shellcheck disable=SC2012
        KERNEL_VERSION="$(cd /lib/modules && ls -1v | tail -1)"
    fi
fi

find_initrd_for_kernel_version() {
    local kernel_version="$1"
    local base_path files initrd machine_id

    if [[ -d /efi/Default ]] || [[ -d /boot/Default ]] || [[ -d /boot/efi/Default ]]; then
        machine_id="Default"
    elif [[ -s /etc/machine-id ]]; then
        read -r machine_id < /etc/machine-id
        [[ $machine_id == "uninitialized" ]] && machine_id="Default"
    else
        machine_id="Default"
    fi

    if [ -n "$machine_id" ]; then
        for base_path in /efi /boot /boot/efi; do
            initrd="${base_path}/${machine_id}/${kernel_version}/initrd"
            if [ -f "$initrd" ]; then
                echo "$initrd"
                return
            fi
        done
    fi

    if [[ -f /lib/modules/${kernel_version}/initrd ]]; then
        echo "/lib/modules/${kernel_version}/initrd"
    elif [[ -f /lib/modules/${kernel_version}/initramfs.img ]]; then
        echo "/lib/modules/${kernel_version}/initramfs.img"
    elif [[ -f /boot/initramfs-${kernel_version}.img ]]; then
        echo "/boot/initramfs-${kernel_version}.img"
    elif [[ -f /usr/lib/modules/${kernel_version}/initramfs.img ]]; then
        echo "/usr/lib/modules/${kernel_version}/initramfs.img"
    else
        files=(/boot/initr*"${kernel_version}"*)
        if [ "${#files[@]}" -ge 1 ] && [ -e "${files[0]}" ]; then
            echo "${files[0]}"
        fi
    fi
}

if [[ $1 ]]; then
    image="$1"
    if ! [[ -f $image ]]; then
        {
            echo "$image does not exist"
            echo
        } >&2
        usage
        exit 1
    fi
else
    image=$(find_initrd_for_kernel_version "$KERNEL_VERSION")
fi

shift
while (($# > 0)); do
    filenames[${1#/}]=1
    shift
done

if ! [[ -f $image ]]; then
    {
        echo "No <initramfs file> specified and the default image '$image' cannot be accessed!"
        echo
    } >&2
    usage
    exit 1
fi
image=$(realpath "$image")

TMPDIR="$(mktemp -d -t lsinitrd.XXXXXX)"
# shellcheck disable=SC2064
trap "rm -rf '$TMPDIR'" EXIT

dracutlibdirs() {
    for d in lib64/dracut lib/dracut usr/lib64/dracut usr/lib/dracut; do
        echo "$d/$1"
    done
}

SQUASH_TMPFILE=""
SQUASH_EXTRACT="$TMPDIR/squash-extract"

# Takes optional pattern arguments
cpio_extract() {
    if [ "$EXTRACTOR" = 3cpio ]; then
        3cpio --extract --make-directories --parts "$parts" $verbose "$image" -- "$@"
    else
        "$EXTRACTOR" --parts "$parts" $verbose "$image" -- "$@"
    fi
}

# Takes optional pattern arguments
cpio_extract_to_stdout() {
    if [ "$EXTRACTOR" = 3cpio ]; then
        3cpio --extract --parts "$parts" --to-stdout "$image" -- "$@"
    else
        "$EXTRACTOR" --parts "$parts" --to-stdout "$image" -- "$@"
    fi
}

# Takes optional pattern arguments
cpio_list() {
    if [ "$EXTRACTOR" = 3cpio ]; then
        3cpio --list --parts "$parts" --verbose "$image" -- "$@"
    else
        "$EXTRACTOR" --list --parts "$parts" --verbose "$image" -- "$@"
    fi
}

extract_squash_img() {
    local _img _tmp

    [[ $SQUASH_TMPDIR == none ]] && return 1
    [[ -s $SQUASH_TMPFILE ]] && return 0

    # Before dracut 104 the image was named squash-root.img. Keep the old name
    # so newer versions of lsinitrd can inspect initrds build with older dracut
    # versions.
    for _img in squash-root.img squashfs-root.img erofs-root.img; do
        _tmp="$TMPDIR/$_img"
        cpio_extract_to_stdout "$_img" > "$_tmp"
        [[ -s $_tmp ]] || continue

        SQUASH_TMPFILE="$_tmp"

        # fsck.erofs doesn't allow extracting single files or listing the
        # content of the image. So always extract the full image.
        if [[ $_img == erofs-root.img ]]; then
            mkdir -p "$SQUASH_EXTRACT"
            fsck.erofs --extract="$SQUASH_EXTRACT/erofs-root" --overwrite "$SQUASH_TMPFILE" 2> /dev/null
            ((ret += $?))
        fi

        break
    done

    if [[ -z $SQUASH_TMPFILE ]]; then
        SQUASH_TMPDIR=none
        return 1
    fi

    return 0
}

extract_files() {
    local nofileinfo

    ((${#filenames[@]} == 1)) && nofileinfo=1
    for f in "${!filenames[@]}"; do
        [[ $nofileinfo ]] || echo "initramfs:/$f"
        [[ $nofileinfo ]] || echo "========================================================================"
        # shellcheck disable=SC2001
        [[ $f == *"\\x"* ]] && f=$(echo "$f" | sed 's/\\x.\{2\}/????/g')

        case $f in
            squashfs-root/*)
                extract_squash_img
                unsquashfs -force -d "$SQUASH_EXTRACT" -no-progress "$SQUASH_TMPFILE" "${f#squashfs-root/}" &> /dev/null
                ((ret += $?))
                cat "$SQUASH_EXTRACT/${f#squashfs-root/}" 2> /dev/null
                ;;
            erofs-root/*)
                extract_squash_img
                cat "$SQUASH_EXTRACT/$f" 2> /dev/null
                ;;
            *)
                cpio_extract_to_stdout "$f"
                ((ret += $?))
                ;;
        esac

        [[ $nofileinfo ]] || echo "========================================================================"
        [[ $nofileinfo ]] || echo
    done
}

list_modules() {
    echo "dracut modules:"
    # shellcheck disable=SC2046
    cpio_extract_to_stdout $(dracutlibdirs modules.txt)
    ((ret += $?))
}

list_files() {
    echo "========================================================================"
    if [ "$sorted" -eq 1 ]; then
        cpio_list | sort -n -k5
    else
        cpio_list
    fi
    ((ret += $?))
    echo "========================================================================"
}

list_squash_content() {
    extract_squash_img || return 0

    echo "Squashed content (${SQUASH_TMPFILE##*/}):"
    echo "========================================================================"
    case $SQUASH_TMPFILE in
        */squash-root.img | */squashfs-root.img)
            unsquashfs -ll "$SQUASH_TMPFILE" | tail -n +4
            ;;
        */erofs-root.img)
            (
                cd "$SQUASH_EXTRACT" || return 1
                find erofs-root/ -exec ls -ld {} +
            )
            ;;
    esac
    echo "========================================================================"
}

list_cmdline() {

    echo "dracut cmdline:"
    cpio_extract_to_stdout etc/cmdline.d/\*.conf
    ((ret += $?))

    extract_squash_img || return 0
    case $SQUASH_TMPFILE in
        */squash-root.img | */squashfs-root.img)
            unsquashfs -force -d "$SQUASH_EXTRACT" -no-progress "$SQUASH_TMPFILE" etc/cmdline.d/\*.conf &> /dev/null
            ((ret += $?))
            cat "$SQUASH_EXTRACT"/etc/cmdline.d/*.conf 2> /dev/null
            ;;
        */erofs-root.img)
            cat "$SQUASH_EXTRACT"/erofs-root/etc/cmdline.d/*.conf 2> /dev/null
            ;;
    esac

}

unpack_files() {
    if ((${#filenames[@]} > 0)); then
        for f in "${!filenames[@]}"; do
            # shellcheck disable=SC2001
            [[ $f == *"\\x"* ]] && f=$(echo "$f" | sed 's/\\x.\{2\}/????/g')
            case $f in
                squashfs-root/*)
                    extract_squash_img || continue
                    unsquashfs -force -d "squashfs-root" -no-progress "$SQUASH_TMPFILE" "${f#squashfs-root/}" > /dev/null
                    ((ret += $?))
                    ;;
                erofs-root/*)
                    extract_squash_img || continue
                    mkdir -p "${f%/*}"
                    cp -rf "$SQUASH_EXTRACT/$f" "$f"
                    ;;
                *)
                    cpio_extract "$f"
                    ((ret += $?))
                    ;;
            esac
        done
    else
        cpio_extract
        ((ret += $?))

        extract_squash_img || return 0
        case $SQUASH_TMPFILE in
            */squash-root.img | */squashfs-root.img)
                unsquashfs -d "squashfs-root" -no-progress "$SQUASH_TMPFILE" > /dev/null
                ((ret += $?))
                ;;
            */erofs-root.img)
                cp -rf "$SQUASH_EXTRACT/erofs-root" .
                ;;
        esac
    fi
}

read -r -N 2 bin < "$image"
if [ "$bin" = "MZ" ]; then
    command -v "${OBJCOPY:-objcopy}" > /dev/null || {
        echo "Need 'objcopy' to unpack an UEFI executable."
        exit 1
    }
    "${OBJCOPY:-objcopy}" \
        --dump-section .linux="$TMPDIR/vmlinuz" \
        --dump-section .initrd="$TMPDIR/initrd.img" \
        --dump-section .cmdline="$TMPDIR/cmdline.txt" \
        --dump-section .osrel="$TMPDIR/osrel.txt" \
        "$image" /dev/null
    uefi="$image"
    image="$TMPDIR/initrd.img"
    [ -f "$image" ] || exit 1
fi

if ((${#filenames[@]} <= 0)) && [[ -z $unpack ]] && [[ -z $unpackearly ]]; then
    if [ -n "$uefi" ]; then
        echo -n "initrd in UEFI: $uefi: "
        du -h "$image" | while read -r a _ || [ -n "$a" ]; do echo "$a"; done
        if [ -f "$TMPDIR/osrel.txt" ]; then
            name=$(sed -En '/^PRETTY_NAME/ s/^\w+=["'"'"']?([^"'"'"'$]*)["'"'"']?/\1/p' "$TMPDIR/osrel.txt")
            id=$(sed -En '/^ID/ s/^\w+=["'"'"']?([^"'"'"'$]*)["'"'"']?/\1/p' "$TMPDIR/osrel.txt")
            build=$(sed -En '/^BUILD_ID/ s/^\w+=["'"'"']?([^"'"'"'$]*)["'"'"']?/\1/p' "$TMPDIR/osrel.txt")
            echo "OS Release: $name (${id}-${build})"
        fi
        if [ -f "$TMPDIR/vmlinuz" ]; then
            version=$(strings -n 20 "$TMPDIR/vmlinuz" | sed -En '/[0-9]+\.[0-9]+\.[0-9]+/ { p; q 0 }')
            echo "Kernel Version: $version"
        fi
        if [ -f "$TMPDIR/cmdline.txt" ]; then
            echo "Command line:"
            sed -En 's/\s+/\n/g; s/\x00/\n/; p' "$TMPDIR/cmdline.txt"
        fi
    else
        echo -n "Image: $image: "
        du -h "$image" | while read -r a _ || [ -n "$a" ]; do echo "$a"; done
    fi

    echo "========================================================================"
fi

unset skip
read -r -N 6 bin < "$image"
case $bin in
    $'\x71\xc7'* | 070701)
        parts=1
        is_early=$(cpio_extract_to_stdout early_cpio 2> /dev/null)
        # Debian mkinitramfs does not create the file 'early_cpio', so let's check if firmware files exist
        [[ "$is_early" ]] || is_early=$(cpio_list 'kernel/*/microcode/*.bin' 2> /dev/null)
        if [[ "$is_early" ]]; then
            if [[ -n $unpack ]]; then
                # should use --unpackearly for early CPIO
                :
            elif [[ -n $unpackearly ]]; then
                unpack_files
            elif ((${#filenames[@]} > 0)); then
                extract_files
            else
                echo "Early CPIO image"
                list_files
            fi
            parts=2-
        else
            parts=1-
        fi
        ;;
    *)
        parts=1-
        ;;
esac

ret=0

if [[ -n $unpack ]]; then
    unpack_files
elif ((${#filenames[@]} > 0)); then
    extract_files
else
    # shellcheck disable=SC2046
    version=$(cpio_extract_to_stdout $(dracutlibdirs 'dracut-*'))
    ((ret += $?))
    echo "Version: $version"
    echo
    if [ "$modules" -eq 1 ]; then
        list_modules
        echo "========================================================================"
    else
        echo -n "Arguments: "
        # shellcheck disable=SC2046
        cpio_extract_to_stdout $(dracutlibdirs build-parameter.txt)
        echo
        list_modules
        list_files
        list_squash_content
        echo
        list_cmdline
    fi
fi

exit "$ret"
